### ISSUE TITLE
about DC_and_HD_loss

### ISSUE THREAD

**Issue by GuobinZhangTJU** [JunMa11/SegLoss#25]
Hi, Jun, I want to use DC+HD as the loss to train a multi-organ segmentation model, but I meet two problems:

1. in _nnUNetTrainer_DiceHDBinary.py: from nnunet.training.loss_functions.boundary_loss_ import **DC_and_HDBinary_loss**, 
    but there is no **DC_and_HDBinary_loss** in _boundary_loss.py_, only has **DC_and_HD_loss**, are they the same?
2. in the defination of **DC_and_HD_loss**
`    def forward(self, net_output, target):
        dc_loss = self.dc(net_output, target)
        hd_loss = self.hd(net_output, target)
        if self.aggregate == "sum":
            with torch.no_grad():
                alpha = hd_loss / (dc_loss + 1e-5 )
            result = alpha * dc_loss + hd_loss
        else:
            raise NotImplementedError("nah son")
        return result  `
Here,
alpha = hd_loss / (dc_loss + 1e-5 ), result = alpha * dc_loss + hd_loss, 
alpha * dc_loss = hd_loss / (dc_loss + 1e-5 )  * dc_loss, where 1e-5 is too small which can be ignore, then the result will be hd_loss , the final result will equals to 2*hd_loss, not hd_loss+dc_loss, isn't it?
Thanks, waiting for your reply~

**Comment by JunMa11** [comment]
Hi @GuobinZhangTJU ,

Sorry for my late reply. I'm quite busy recently.

I used this implementation

https://github.com/JunMa11/SegLoss/blob/8838616323b6ffa7fdb522c7d0a5250be4610221/test/loss_functions/boundary_loss.py#L286

I also upload the trainer https://github.com/JunMa11/SegLoss/blob/master/test/network_training/nnUNetTrainer_DiceHD.py

Tip: I would recommend to first train your network with Dice loss and then finetune with DiceHD loss.

Best,
Jun

**Comment by GuobinZhangTJU** [comment]
> Hi @GuobinZhangTJU ,
> 
> Sorry for my late reply. I'm quite busy recently.
> 
> I used this implementation
> 
> https://github.com/JunMa11/SegLoss/blob/8838616323b6ffa7fdb522c7d0a5250be4610221/test/loss_functions/boundary_loss.py#L286
> 
> I also upload the trainer https://github.com/JunMa11/SegLoss/blob/master/test/network_training/nnUNetTrainer_DiceHD.py
> 
> Tip: I would recommend to first train your network with Dice loss and then finetune with DiceHD loss.
> 
> Best,
> Jun

Many thanks, Jun, wish the work is great~~

**Comment by GuobinZhangTJU** [comment]
> Hi @GuobinZhangTJU ,
> 
> Sorry for my late reply. I'm quite busy recently.
> 
> I used this implementation
> 
> https://github.com/JunMa11/SegLoss/blob/8838616323b6ffa7fdb522c7d0a5250be4610221/test/loss_functions/boundary_loss.py#L286
> 
> I also upload the trainer https://github.com/JunMa11/SegLoss/blob/master/test/network_training/nnUNetTrainer_DiceHD.py
> 
> Tip: I would recommend to first train your network with Dice loss and then finetune with DiceHD loss.
> 
> Best,
> Jun

```
class DC_and_HD_loss(nn.Module):
    def __init__(self, soft_dice_kwargs, hd_kwargs, aggregate="sum"):
        super(DC_and_HD_loss, self).__init__()
        self.aggregate = aggregate
        self.dc = SoftDiceLoss(apply_nonlin=softmax_helper, **soft_dice_kwargs)
        self.hd = HDLoss(**hd_kwargs)

    def forward(self, net_output, target):
        dc_loss = self.dc(net_output, target)
        hd_loss = self.hd(net_output, target)
        if self.aggregate == "sum":
            with torch.no_grad():
                alpha = hd_loss / (dc_loss + 1e-5 )   ##  Here 1
            result = alpha * dc_loss + hd_loss  ##  Here 2
        else:
            raise NotImplementedError("nah son")
        return result 
```

Hi, Jun, in ##  Here 1 and ##  Here 2,
alpha = hd_loss / (dc_loss + 1e-5 ), 
result = alpha * dc_loss + hd_loss,
alpha * dc_loss = hd_loss / (dc_loss + 1e-5 ) * dc_loss, where 1e-5 is too small which can be ignore, then the result will be 
hd_loss , the final result will equals to 2*hd_loss, not hd_loss+dc_loss, isn't it?

**Comment by JunMa11** [comment]
@GuobinZhangTJU ,

The direct answer is no.

alpha is a scalar, which does not participate in BP and can not be merged in `Here 2`. 
In other words, it is  just a weight.

**Comment by GuobinZhangTJU** [comment]
> @GuobinZhangTJU ,
> 
> The direct answer is no.
> 
> alpha is a scalar, which does not participate in BP and can not be merged in `Here 2`.
> In other words, it is just a weight.

Got it~  Thanks~~Jun~
